### PR TITLE
Add subdirectory option

### LINKED ISSUE #183: Use a subdirectory
Often times a template is wrapped in a subdirectory of your repository, because there are files you do not want to include into the template, or scripts for updating contents, etc.

Examples:
  - README: you want to explain what is your template, how to generate a project from it, how to use the generated project, etc. This does not belong to the template itself
  - LICENSE: your template is licensed, but the user might not want to use the same license in the generated project.

So my question is: is it possible to render only a subdirectory of the git repository specified as source?

I tried defining `_src_path` in `copiers.yml` but it had no effect.

### PR DESCRIPTION
**Pull request by pawamoy**
Well, this required less code than I thought! I guess I could add more tests, particularly when updating a project.

TODO:
- [x] docs
- [x] test when updating

Fixes #183

### CODE REVIEW

**Comment by pawamoy** [comment]
It would be great if you could review and merge #192 so I can rebase before adding documentation for this feature!

File: copier/cli.py [hunk 1]
```diff
@@ -75,6 +75,15 @@ class CopierApp(cli.Application):
             "the latest version, use `--vcs-ref=HEAD`."
         ),
     )
+    subdirectory = cli.SwitchAttr(
+        ["-D", "--subdirectory"],
```
Comment (pawamoy) [inline copier/cli.py:79]: The `-D` option is already used by the `-D, --no-diff` option of the `update` subcommand, and I'm not sure both can live together. Any better idea than `D`? `s` and `d` are already taken. Other letters like `u`, `y` do not feel right. Maybe `b`? Or capital `S`?

Comment (Yajo) [reply copier/cli.py:79]: Both can coexist, but I prefer to choose a different one.

`b` is good IMHO.

File: tests/test_subdirectory.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from plumbum import local
+from plumbum.cmd import git
+
+import copier
+from copier.config import make_config
+from copier.main import update_diff
+
+
+def test_copy_subdirectory_api_option(dst):
```
Comment (Yajo) [inline tests/test_subdirectory.py:9]: I plan to deprectate `dst` as pylint has a valid alternative fixture upstream:

```suggestion
def test_copy_subdirectory_api_option(tmp_path):
```

The same in the rest of your test, please.

**Comment by Yajo** [comment]
This should be a WIP until #198 is fixed IMHO.

**Comment by pawamoy** [comment]
> 1.  Template in v1 doesn't have a _subdirectory; in v2 it moves all things into a subdir and adds that key to copier.yml. Some files change. Downstream project has evolved too. Does that work as expected?
> 2. Template in v3 changes from one subdirectory to another. Some file evolves also. Subproject evolves separately. Subproject is updated. Does that work as expected?

These are great remarks, thank you for having taken the time to think about this. I'll add tests to make sure what you listed is working as expected :slightly_smiling_face:

**Comment by pawamoy** [comment]
I've added two *big* tests with comments :slightly_smiling_face: Let me know what you think!

File: README.md [hunk 3]
```diff
@@ -547,7 +552,9 @@ copier.copy(
     force=False,
     skip=False,
     quiet=False,
-    cleanup_on_error=True
+    cleanup_on_error=True,
+
+    subdirectory=None
```
Comment (Yajo) [inline README.md:557]:
```suggestion
    subdirectory=None,
```

File: README.md [hunk 4]
```diff
@@ -623,14 +630,17 @@ Uses the template in _src_path_ to generate a new project at _dst_path_.
...
+
 ## Comparison with other project generators
 
 ### Cookiecutter
 
 Cookiecutter and Copier are quite similar in functionality, except that:
 
-- Cookiecutter uses a subdirectory to generate the project, while Copier uses the root
-  directory.
+- Cookiecutter uses a subdirectory to generate the project, while Copier can use both
```
Comment (Yajo) [inline README.md:642]:
```suggestion
- Cookiecutter uses a subdirectory to generate the project, while Copier can use either
```

File: tests/test_copy.py [hunk 5]
```diff
@@ -153,3 +153,11 @@ def test_pretend_option(dst):
...
     assert not (dst / "config.py").exists()
     assert not (dst / "pyproject.toml").exists()
+
+
+def test_subdirectory(dst):
+    render(dst, subdirectory="doc")
+    assert not (dst / "doc").exists()
+    assert not (dst / "config.py").exists()
+    assert (dst / "images").exists()
+    assert (dst / "manana.txt").exists()
```
Comment (Yajo) [inline tests/test_copy.py:163]: I want to remove the `dst` fixture at some point. Please don't use it for new code:

```suggestion
def test_subdirectory(tmp_path: Path):
    render(tmp_path, subdirectory="doc")
    assert not (tmp_path / "doc").exists()
    assert not (tmp_path / "config.py").exists()
    assert (tmp_path / "images").exists()
    assert (tmp_path / "manana.txt").exists()
```

File: tests/test_subdirectory.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    git("config", "user.email", "test@copier")
+    git("add", ".")
+    git("commit", "-m", message)
+
+
+def test_copy_subdirectory_api_option(tmp_path):
+    copier.copy(
+        "./tests/demo_subdirectory", tmp_path, force=True, subdirectory="api_project"
+    )
+    assert (tmp_path / "api_readme.md").exists()
```
Comment (Yajo) [inline tests/test_subdirectory.py:23]:
```suggestion
    assert (tmp_path / "api_readme.md").exists()
    assert not (tmp_path / "conf_readme.md").exists()
```

File: tests/test_subdirectory.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+
+def test_copy_subdirectory_config(tmp_path):
+    copier.copy("./tests/demo_subdirectory", tmp_path, force=True)
+    assert (tmp_path / "conf_readme.md").exists()
```
Comment (Yajo) [inline tests/test_subdirectory.py:28]:
```suggestion
    assert (tmp_path / "conf_readme.md").exists()
    assert not (tmp_path / "api_readme.md").exists()
```

File: tests/test_subdirectory.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+    # First, create the template with an initial README
+    with local.cwd(template_path):
+        with open("README.md", "w") as fd:
+            fd.write("upstream version 1\n")
+
+        with open("[[_copier_conf.answers_file]].tmpl", "w") as fd:
+            fd.write("[[_copier_answers|to_nice_yaml]]\n")
+
+        git_init("hello template")
```
Comment (Yajo) [inline tests/test_subdirectory.py:60]: To properly assert project history evolution works, add tags:

```suggestion
        git_init("hello template")
        git("tag", "v1")
```

File: tests/test_subdirectory.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+    # Generate the project a first time, assert the README exists
+    copier.copy(str(template_path), project_path, force=True)
+    assert (project_path / "README.md").exists()
```
Comment (Yajo) [inline tests/test_subdirectory.py:64]: Then check those are used:

```suggestion
    assert (project_path / "README.md").exists()
    assert "_commit: v1" in (project_path / ".copier-answers.yml").read_text()
```

File: tests/test_subdirectory.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+            "subdir/[[_copier_conf.answers_file]].tmpl",
+        )
+
+        # Add the subdirectory option to copier.yml
+        with open("copier.yml", "w") as fd:
+            fd.write("_subdirectory: subdir\n")
+
+        # Commit the changes
+        git("add", ".", "-A")
+        git("commit", "-m", "use a subdirectory now")
```
Comment (Yajo) [inline tests/test_subdirectory.py:96]:
```suggestion
        git("commit", "-m", "use a subdirectory now")
        git("tag", "v2")
```

File: tests/test_subdirectory.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+    # Finally, update the generated project
+    copier.copy(str(template_path), project_path, force=True)
```
Comment (Yajo) [inline tests/test_subdirectory.py:99]: `src` should be obtained automatically from `.copier-answers.yml`

```suggestion
    copier.copy(dst=project_path, force=True)
    assert "_commit: v2" in (project_path / ".copier-answers.yml").read_text()
```

File: tests/test_copy.py (hunk 5 above)
Comment (pawamoy) [reply tests/test_copy.py:163]: Woops, sure, sorry, missed that one

File: tests/test_subdirectory.py (hunk 11 above)
Comment (pawamoy) [reply tests/test_subdirectory.py:104]: Thanks, I wasn't sure how to do this correctly.

**Comment by pawamoy** [comment]
Hey @Yajo, not sure if you were waiting for me to comment, but I implemented the changes you suggested :slightly_smiling_face: 

If you just couldn't find the time to review it, then please ignore this comment, there is absolutely nothing urgent here!

**Comment by Yajo** [comment]
Yes, I was waiting. Thanks for your contribution!

**Comment by pawamoy** [comment]
Would it be possible for you to release a new version with this new feature @Yajo? That would be great 🙂

**Comment by Yajo** [comment]
Oh yes, I was hoping to have time to put more features, but I've been lacking time lately.

Let me release.

**Comment by micheelengronne** [comment]
The `_subdirectory` parameter does not seem to be respected when doing `copier update`.

I put it in my `.copier-answers.yml` but copier wants to remove it and fetch the repository root anyway.

**Comment by Yajo** [comment]
https://copier.readthedocs.io/en/latest/updating/#never-change-the-answers-file-manually

**Comment by micheelengronne** [comment]
Yes, but the subdirectory parameter was also set on the `copier.yml` file and was no retain by the answers file.

**Comment by micheelengronne** [comment]
By the way, I think in a CI system, bots should update the answers file to update `_commit` for instance and keep the template system immutable.

**Comment by Yajo** [comment]
Hi there @micheelengronne, according to your comments I guess that you're having some trouble understanding what this option does and how Copier updates your subprojects.

If you want, you can open up a new [discussion](https://github.com/copier-org/copier/discussions) explaining your use case so you can get better help from the community. This thread was finished long ago and is not the best place to get further support.

**Comment by micheelengronne** [comment]
Sorry @Yajo Indeed, I read too fast that part of the doc. I altered my code accordingly and I will open a new issue with my ideas and a snippet.

**Comment by micheelengronne** [comment]
I will delete my previous comments. They are totally out of subject.

**Pull request merged**